Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp

```diff
@@ -348,288 +338,314 @@ PlatformLayer* AudioVideoRendererRemote::platformVideoLayer() const
+    ensureOnDispatcherWithConnection([trackIdentifier](auto& renderer, auto& connection) {
+        connection.send(Messages::RemoteAudioVideoRendererProxyManager::RemoveTrack(renderer.m_identifier, trackIdentifier), 0);
+    });
 }
 
 void AudioVideoRendererRemote::enqueueSample(TrackIdentifier trackIdentifier, Ref<MediaSample>&& sample, std::optional<MediaTime> expectedMinimum)
 {
-    RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
-    if (!isGPURunning() || !gpuProcessConnection)
-        return;
-    gpuProcessConnection->connection().send(Messages::RemoteAudioVideoRendererProxyManager::EnqueueSample(m_identifier, trackIdentifier, MediaSamplesBlock::fromMediaSample(sample), expectedMinimum), 0);
-    if (auto it = m_requestMediaDataWhenReadyData.find(trackIdentifier); it != m_requestMediaDataWhenReadyData.end())
-        it->value.pendingSamples++;
+    {
+        Locker locker { m_lock };
+        readyForMoreData(trackIdentifier).pendingSamples++;
+    }
+    ensureOnDispatcherWithConnection([trackIdentifier, sample = WTFMove(sample), expectedMinimum](auto& renderer, auto& connection) {
+        connection.send(Messages::RemoteAudioVideoRendererProxyManager::EnqueueSample(renderer.m_identifier, trackIdentifier, MediaSamplesBlock::fromMediaSample(sample), expectedMinimum), 0);
+    });
 }
 
 bool AudioVideoRendererRemote::isReadyForMoreSamples(TrackIdentifier trackIdentifier)
 {
-    auto it = m_requestMediaDataWhenReadyData.find(trackIdentifier);
-    return it != m_requestMediaDataWhenReadyData.end() && it->value.readyForMoreData();
+    Locker locker { m_lock };
+    return readyForMoreData(trackIdentifier).isReadyForMoreData();
 }
 
 void AudioVideoRendererRemote::requestMediaDataWhenReady(TrackIdentifier trackIdentifier, Function<void(TrackIdentifier)>&& callback)
 {
-    RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
-    if (!isGPURunning() || !gpuProcessConnection)
-        return;
-    if (auto it = m_requestMediaDataWhenReadyData.find(trackIdentifier); it != m_requestMediaDataWhenReadyData.end())
-        it->value.callback = WTFMove(callback);
-    else {
-        RequestMediaDataWhenReadyData data { .callback = WTFMove(callback) };
-        m_requestMediaDataWhenReadyData.set(trackIdentifier, WTFMove(data));
-    }
-    gpuProcessConnection->connection().send(Messages::RemoteAudioVideoRendererProxyManager::RequestMediaDataWhenReady(m_identifier, trackIdentifier), 0);
+    ensureOnDispatcherWithConnection([trackIdentifier, callback = WTFMove(callback)](auto& renderer, auto& connection) mutable {
+        assertIsCurrent(queueSingleton());
+        if (auto it = renderer.m_requestMediaDataWhenReadyDataCallbacks.find(trackIdentifier); it != renderer.m_requestMediaDataWhenReadyDataCallbacks.end())
+            it->value = WTFMove(callback);
+        else
+            renderer.m_requestMediaDataWhenReadyDataCallbacks.set(trackIdentifier, WTFMove(callback));
```

Comment:
Why doing potentially two finds?
Shouldn't we just do a set here?

Reply:
no because I want to keep the current counter